src/iosrv/sched.c

```diff
@@ -0,0 +1,542 @@
+ * portions thereof marked with this legend must also reproduce the markings.
+ */
+
+#define D_LOGFAC       DD_FAC(server)
+
+#include <abt.h>
+#include <daos/common.h>
+#include <daos_errno.h>
+#include "srv_internal.h"
+
+/*
+ * A schedule cycle consists of three stages:
+ * 1. Starting with a network poll ULT, number of ULTs to be executed in this
+ *    cycle is queried by ABT_pool_get_size() for each non-poll ABT pool;
+ * 2. Executing all other ULTs which not for hardware polling;
+ * 3. Ending with a NVMe poll ULT;
+ *
+ * Extra network & NVMe poll ULTs could be scheduled in executing stage
+ * according to network/NVMe poll age and request/IO statistics.
+ */
+struct sched_cycle {
+	uint32_t	sc_ults_cnt[DSS_POOL_CNT];
+	uint32_t	sc_ults_tot;
+	/*
+	 * bound[0]: Minimum network/NVMe poll age, scheduler always try to
+	 * execute few ULTs (if there is any) before next poll.
+	 *
+	 * bound[1]: Maximum network/NVMe poll age, scheduler will do an extra
+	 * poll if it's not polled after executing cerntain amout of ULTs.
+	 */
+	uint32_t	sc_age_net_bound[2];
+	uint32_t	sc_age_nvme_bound[2];
+	uint32_t	sc_age_net;
+	uint32_t	sc_age_nvme;
+	unsigned int	sc_new_cycle:1,
+			sc_cycle_started:1;
+};
+
+struct sched_data {
+	struct sched_cycle	 sd_cycle;
+	struct dss_xstream	*sd_dx;
+	uint32_t		 sd_event_freq;
+};
+
+static unsigned int sched_throttle[DSS_POOL_CNT] = {
+	0,	/* DSS_POOL_NET_POLL */
+	0,	/* DSS_POOL_NVME_POLL */
+	0,	/* DSS_POOL_IO */
+	30,	/* DSS_POOL_REBUILD */
+	0,	/* DSS_POOL_AGGREGATE */
+	0,	/* DSS_POOL_GC */
+};
+
+int
+sched_set_throttle(int pool_idx, unsigned int percent)
+{
+	if (percent >= 100) {
+		D_ERROR("Invalid throttle number: %d\n", percent);
+		return -DER_INVAL;
+	}
+
+	if (pool_idx >= DSS_POOL_CNT) {
+		D_ERROR("Invalid pool idx: %d\n", pool_idx);
+		return -DER_INVAL;
+	}
+
+	if (pool_idx == DSS_POOL_NET_POLL || pool_idx == DSS_POOL_NVME_POLL) {
+		D_ERROR("Can't throttle network or NVMe poll\n");
+		return -DER_INVAL;
+	}
+
+	sched_throttle[pool_idx] = percent;
+	return 0;
+}
+
+/* #define SCHED_DEBUG */
+static void
+sched_dump_data(struct sched_data *data)
+{
+#ifdef SCHED_DEBUG
+	struct dss_xstream	*dx = data->sd_dx;
+	struct sched_cycle	*cycle = &data->sd_cycle;
+
+	D_PRINT("XS(%d): comm:%d main:%d. age_net:%u, [%u, %u], "
+		"age_nvme:%u, [%u, %u] new_cycle:%d cycle_started:%d "
+		"total_ults:%u [%u, %u, %u, %u]\n", dx->dx_xs_id,
+		dx->dx_comm, dx->dx_main_xs, cycle->sc_age_net,
+		cycle->sc_age_net_bound[0], cycle->sc_age_net_bound[1],
+		cycle->sc_age_nvme, cycle->sc_age_nvme_bound[0],
+		cycle->sc_age_nvme_bound[1], cycle->sc_new_cycle,
+		cycle->sc_cycle_started, cycle->sc_ults_tot,
+		cycle->sc_ults_cnt[DSS_POOL_IO],
+		cycle->sc_ults_cnt[DSS_POOL_REBUILD],
+		cycle->sc_ults_cnt[DSS_POOL_AGGREGATE],
+		cycle->sc_ults_cnt[DSS_POOL_GC]);
+#endif
+}
+
+#define SCHED_AGE_NET_MIN		32
+#define SCHED_AGE_NET_MAX		512
+#define SCHED_AGE_NVME_MIN		32
+#define SCHED_AGE_NVME_MAX		512
```

Comment:
These are just some random number? or there are reasons behind?

Reply:
random numbers so far, it can be tuned later.

Reply:
so you mean SCHED_AGE_NET_MAX should be smaller? 'MIN' value is actually used to avoid calling poll function too often.